### PR TITLE
Add video play for windows guest

### PR DESCRIPTION
**Pull request by suqinhuang**
Play video with wmplayer if it's installed default,
else install kmplayer and run video

Signed-off-by: Suqin Huang <shuang@redhat.com>

id:1403637

### CODE REVIEW

File: qemu/tests/win_video_play.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    def kmplayer_install(session, vm, tmp_dir):
+        """
+        Install kmplayer
+        """
+        error_context.context("Install kmplayer ...", logging.info)
+        guest_name = params["guest_name"]
+        alias_map = params.get("guest_alias")
+        if alias_map:
+            guest_list = dict([x.split(":") for x in alias_map.split(";")])
+            print guest_list
```
Comment (PyLearner) [inline qemu/tests/win_video_play.py]: is this used to debug ?

File: qemu/tests/win_video_play.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                                           "pls check cfg file")
+            else:
+                folder_name = guest_list[guest_name].split(",")[0]
+                pkg_md5sum = guest_list[guest_name].split(",")[1]
+        src_url = params["kmplayer_source"] % folder_name
+        error_context.context("Download kmplayer source files", logging.info)
+        pkg_tmp = utils.unmap_url_cache(tmp_dir, src_url, pkg_md5sum)
+        error_context.context("Extract the pkg...", logging.info)
+        tar_cmd = params.get("tar_cmd", "tar xvf")
+        tar_cmd = "%s %s -C %s" % (tar_cmd, pkg_tmp, tmp_dir)
```
Comment (PyLearner) [inline qemu/tests/win_video_play.py]: import tarfile
t = tarfile.open('/path/pkg.tar', 'r:')
if it is gz file, could t = tarfile.open('/path/pkg.tar', 'r:gz')
t.extractall(path='extract_dir')
t.close
does this module meet your demand

**Comment by suqinhuang** [comment]
the kmplayer files is around 1G, it's better to put in web instead winutils.iso

File: qemu/tests/cfg/win_video_play.cfg [hunk 3]
```diff
@@ -0,0 +1,17 @@
+- win_video_play:
+    type = win_video_play
+    only Windows
+    #Install kmplayer if wmplayer is not installed default
+    kmplayer_path = "C:\KMPlayer\kmplayer.exe"
+    kmplayer_source =
```
Comment (PyLearner) [inline qemu/tests/cfg/win_video_play.cfg]: seems this parameter needs a value.

Comment (suqinhuang) [inline qemu/tests/cfg/win_video_play.cfg]: send it internal

File: qemu/tests/win_video_play.py (hunk 2 above)
Comment (suqinhuang) [inline qemu/tests/win_video_play.py]: thanks :)

**Comment by xutian** [comment]
please packet the application and install it from software_install.bat

**Comment by CongLi** [comment]
@suqinhuang 
Test failed win win2008-32 guest, could you please help check it?

02:21:06 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/win_video_play.py", line 32, in check_kmplayer_installed
02:21:06 ERROR|     kmplayer_install(session)
02:21:06 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/win_video_play.py", line 47, in kmplayer_install
02:21:06 ERROR|     s, o = session.cmd_status_output(install_cmd)
02:21:06 ERROR|   File "/usr/lib/python2.6/site-packages/aexpect/client.py", line 1136, in cmd_status_output
02:21:06 ERROR|     o = self.cmd_output(cmd, timeout, internal_timeout, print_func, safe)
02:21:06 ERROR|   File "/usr/lib/python2.6/site-packages/aexpect/client.py", line 1057, in cmd_output
02:21:06 ERROR|     raise ShellTimeoutError(cmd, o)
02:21:06 ERROR| ShellTimeoutError: Timeout expired while waiting for shell command to complete: 'start /wait D:\\kmplayer\\2k8\\x86\\KMPlayer-setup.exe /SP- /VERYSILENT'    (output: '')

File: qemu/tests/win_video_play.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        if params.get("kmplayer_reg_cmd"):
+            error_context.context("Update register", logging.info)
+            session.cmd(params.get("kmplayer_reg_cmd"))
+            time.sleep(5)
+
+    play_video_cmd = params["play_video_cmd"] % video_player
+    error_context.context("Play video", logging.info)
+    try:
+        session.cmd(play_video_cmd, timeout=300)
+        time.sleep(params.get("time_for_video"), 240)
```
Comment (CongLi) [inline qemu/tests/win_video_play.py]: This is a typo for the default value.

File: qemu/tests/win_video_play.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+from virttest import error_context
+from virttest import utils_misc
+
+
+@error_context.context_aware
+def run(test, params, env):
+
+    def check_wmplayer_installed(session):
+        """
+        Check if wmplayer is installed
```
Comment (CongLi) [inline qemu/tests/win_video_play.py]: Please add doc string for return value.

File: qemu/tests/win_video_play.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        error_context.context("Check if wmplayer is installed", logging.info)
+        install_path = params.get("wmplayer_path")
+        check_cmd = 'dir "%s"|findstr /I wmplayer'
+        check_cmd = params.get("wmplayer_check_cmd", check_cmd) % install_path
+        if session.cmd_status(check_cmd) == 0:
+            return install_path
+
+    def check_kmplayer_installed(session):
+        """
+        Check if kmplayer is installed
```
Comment (CongLi) [inline qemu/tests/win_video_play.py]: Please add doc string for return value.

File: qemu/tests/win_video_play.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            session.cmd(params.get("kmplayer_reg_cmd"))
+            time.sleep(5)
+
+    play_video_cmd = params["play_video_cmd"] % video_player
+    error_context.context("Play video", logging.info)
+    try:
+        session.cmd(play_video_cmd, timeout=300)
+        time.sleep(params.get("time_for_video"), 240)
+    except Exception, details:
+        raise exceptions.TestError(details)
```
Comment (CongLi) [inline qemu/tests/win_video_play.py]: Should be TestFail?

**Comment by CongLi** [comment]
The access is denied, but the case is pass.
03:54:09 DEBUG| Sending command: "%ProgramFiles%\KMPlayer\kmplayer.exe"  /play /fullscreen
03:54:09 DEBUG| Sending command: echo %errorlevel%
2016-12-30 03:54:09: C:\>
2016-12-30 03:58:09: C:\Program Files\KMPlayer\KMPlayer.exe
2016-12-30 03:58:09: Access is denied.

03:58:32 INFO | PASS 1-smp_8.16384m.repeat1.run_test.Host_RHEL.m6.u9.qcow2.virtio_blk.up.virtio_net.Win2008.i386.sp2.io-github-autotest-qemu.win_video_play

**Comment by CongLi** [comment]
And from the screendump, it stuck at "You must restart your computer after these changes", I guess it was prompted by "registry", need restart guest after registry.

But the test is pass.

Could you please check it again?

**Comment by suqinhuang** [comment]
winutils.iso is updated, it should work now

**Comment by suqinhuang** [comment]
@CongLi 
1.  "You must restart your computer after these changes" is not caused by kmplayer, the prompt shown when I boot up win2008 guest and do nothing.
2. the restart prompt doesn't effect the testing
3. video didn't play because video_url is not provide, I provide a example url now, but need to apply internal patch

**Comment by suqinhuang** [comment]
@xutian @CongLi patch is updated, could you review again

File: qemu/tests/cfg/win_video_play.cfg [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    wmplayer_reg_cmd = "reg add HKLM\SOFTWARE\Policies\Microsoft\WindowsMediaPlayer /v GroupPrivacyAcceptance  /t REG_DWORD /f /d 00000001"
+    wmplayer_path = "${program_files}\Windows Media Player\wmplayer.exe"
+    #Install kmplayer if wmplayer is not installed default
+    kmplayer_install_cmd = "start /wait WIN_UTILS:\kmplayer\%s\KMPlayer-setup.exe /SP- /VERYSILENT"
+    kmplayer_path = "${program_files}\KMPlayer\kmplayer.exe"
+    #Disable IEHarden for kmplayer
+    kmplayer_reg_cmd = reg add "HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Internet Settings\ZoneMap" /v IEHarden /t REG_DWORD /f /d 00000000 &
+    video_url = http://fileshare.com/Peppa_Pig_39_The_Tree_House.avi
+    play_video_cmd = '"%s" ${video_url} /play /fullscreen'
+    guest_alias = "Win2008-sp2-32:2k8\x86,Win2008-sp2-64:2k8\amd64,Win2008-r2-64:2k8\amd64,Win2012-64:2k12\amd64,Win2012-64r2:2k12\amd64"
```
Comment (CongLi) [inline qemu/tests/cfg/win_video_play.cfg]: Why no win10 ?

Comment (suqinhuang) [inline qemu/tests/cfg/win_video_play.cfg]: Win10 is Desktop, mplayer is installed default, and mplayer is also installed in Win2016

Comment (CongLi) [inline qemu/tests/cfg/win_video_play.cfg]: but win10 does not work now :(

**Comment by suqinhuang** [comment]
@CongLi how ahout your testing, it works for me

**Comment by suqinhuang** [comment]
@xutian patch have been updated, could you review it

**Comment by kimi1978wy** [comment]
ran on win2012r2(kmplayer) and win2016(wmplayer) it could pass with out error and play video successfully.
 (1/1) smp_16.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win2012.x86_64.r2.io-github-autotest-qemu.win_video_play: PASS (398.40 s)
 (1/1) smp_16.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.win_video_play: PASS (332.05 s)

File: qemu/tests/win_video_play.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    play_video_cmd = params["play_video_cmd"] % (video_player, video_url)
+    error_context.context("Play video", logging.info)
+    try:
+        session.cmd(play_video_cmd, timeout=300)
+        time.sleep(params.get("time_for_video", 240))
+    except Exception, details:
+        raise exceptions.TestFail(details)
+    finally:
+        error_context.context("Stop video", logging.info)
+        session.cmd('del /f /s "%s"' % video_player)
```
Comment (CongLi) [inline qemu/tests/win_video_play.py]: Need ignore the errors in finally which will cover the real failed reason in case.

Comment (suqinhuang) [inline qemu/tests/win_video_play.py]: I have raise the exception, it should be ok

Comment (CongLi) [inline qemu/tests/win_video_play.py]: But there is finally after exception.

**Comment by suqinhuang** [comment]
PASS 1-smp_16.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win2008.i386.sp2.io-github-autotest-qemu.win_video_play

PASS 2-smp_16.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win2008.x86_64.sp2.io-github-autotest-qemu.win_video_play

PASS 3-smp_16.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win2012.x86_64.io-github-autotest-qemu.win_video_play

File: qemu/tests/win_video_play.py (hunk 9 above)
Comment (CongLi) [inline qemu/tests/win_video_play.py]: 1. I tried to raise error in "try".
1.1 I have updated timeout to 1 in "session.cmd(play_video_cmd, timeout=300)", test pass.
1.2 I have updated the video_url to a non-existing one, test pass.
2. I tried to raise error in "finally", just updated "del" to "el" which is a wrong cmd, test failed, I think such errors should be ignored as this is not the key point.
700 21:09:18 ERROR| Traceback (most recent call last):
701 21:09:18 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-44.0-py2.7.egg/avocado_vt/test.py", line 196, in runTest
702 21:09:18 ERROR|     self._runTest()
703 21:09:18 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-44.0-py2.7.egg/avocado_vt/test.py", line 337, in _runTest
704 21:09:18 ERROR|     run_func(self, params, env)
705 21:09:18 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-44.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
706 21:09:18 ERROR|     return fn(*args, **kwargs)
707 21:09:18 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/win_video_play.py", l    ine 80, in run
708 21:09:18 ERROR|     session.cmd('el /f /s "%s"' % video_player)
709 21:09:18 ERROR|   File "/usr/lib/python2.7/site-packages/aexpect/client.py", line 1211, in cmd
710 21:09:18 ERROR|     raise ShellCmdError(cmd, s, o)
711 21:09:18 ERROR| ShellCmdError: Shell command failed: 'el /f /s "%ProgramFiles%\\Windows Media Player\\wmplayer.exe"'    (status: 9009,    outp    ut: "'el' is not recognized as an internal or external command,\noperable program or batch file.\n\n")
712 21:09:18 ERROR|
713 21:09:18 ERROR| Traceback (most recent call last):
714 
715 21:09:18 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-44.0-py2.7.egg/avocado_vt/test.py", line 434, in _run_avocado
716     raise test_exception
717 
718 21:09:18 ERROR| ShellCmdError: Shell command failed: 'el /f /s "%ProgramFiles%\\Windows Media Player\\wmplayer.exe"'    (status: 9009,    outp    ut: "'el' is not recognized as an internal or external command,\noperable program or batch file.\n\n")

Anyway, as I said, this is an enhancement, it up to you.

**Pull request merged**